### PR TITLE
added basic documentation with example

### PR DESCRIPTION
**Pull request by domdorn**
Basic documentation for understanding ZIO Schema and getting started with a simple Schema + Encoder + Decoder roundtrip.

### CODE REVIEW

**Comment by domdorn** [comment]
My next steps are:
 - document refactoring the schema + fields to be in the companion object
 - show how fields can be used like optics (? I hope they can?)
 - show how to do transformations, e.g. `PersonDTO(firstname: String, lastname: String, age: Int)` => `Person(name: String, age: Int)`

File: docs/overview/index.md [hunk 1]
```diff
@@ -3,8 +3,52 @@ id: overview_index
...
+
+Once you have a representation of your data types, you can use it to 
+ - serialize and deserialize your types
+ - validate your types
+ - transform your types
+ - create instances of your types
+
+You can then use one of the various codecs (or create your own) to serialize and deserialize your types.
+
+Example of codecs are:
```
Comment (thinkharderdev) [inline docs/overview/index.md:25]: Here I would specify which codecs are already implemented vs which ones are possible. Currently we have protobuf and json codecs provided (with Thrift and Avro on the way).

File: docs/overview/first_schema.md [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+
+  val schemaPerson: Schema[Person] = DeriveSchema.gen[Person]
+
+  val schemaPaymentMethod: Schema[PaymentMethod] = DeriveSchema.gen[PaymentMethod]
+ 
+  val schemaCustomer: Schema[Customer] = DeriveSchema.gen[Customer]
+}
+```
+
+## Applying it to our domain
```
Comment (thinkharderdev) [inline docs/overview/first_schema.md:107]: We recently moved codecs into their own modules so I would call out here that you would need to add additional dependencies for these codecs:
```
"dev.zio" %% "zio-schema-json" % "<version>",
"dev.zio" %% "zio-schema-protobuf" % "<version>"
```

File: docs/overview/first_schema.md [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+    _ <- ZIO.debug("is old person the new person? " + (person == newPerson).toString)
+    _ <- ZIO.debug("old person: " + person)
+    _ <- ZIO.debug("new person: " + newPerson)
+  } yield ExitCode.success
+}
+```
+
+
+### Combining different encoders
+Lets take a look at a roundtrip converting an object to JSON and back, then converting it to a protobuf and back.
```
Comment (thinkharderdev) [inline docs/overview/first_schema.md:165]: Small nitpick: Lets should be Let's

File: docs/overview/understanding.md [hunk 4]
```diff
@@ -0,0 +1,328 @@
+---
+id: understanding
+title: "Understanding ZIO-Schema"
+---
+
+
+## Understanding ZIO-Schema
+
+ZIO-Schema is a part of ZIO that is used in many projects (e.g. ZIO-Flow, ZIO-SQL, etc.).
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:9]: I would suggest rewording this as "ZIO Schema is a library used in many ZIO projects such as ZIO Flow and ZIO SQL"

File: docs/overview/understanding.md [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+val age  = person.age
+```
+
+However, this is a "special language feature", it's not "real" like the side effects we've seen in the previous example ( `println(..) vs. Task`println(...)))` ).
+
+Because these basic operations are not "real", we're unable to create an operator that we can use to
+e.g. combine two fields that are inside a nested structure.
+
+The solution to this kind of problem is called an "Optic". Optics provide a way to access the fields of a case class and nested structures.
+There are two main types of optics:
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:55]: Should be "three main types of optics"?

File: docs/overview/understanding.md [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+- `Lens`: A lens is a way to access a field of a case class.
+- `Prims`: A prim is a way to access a field of a nested structure or a collection.
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:57]: Prims -> Prism. I would reword the description a bit. A prism is an optic that allows you to access an instance of a sealed trait as one of its concrete subtypes.

File: docs/overview/understanding.md [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+
+Not only for case classes, but also for other types like collections, tuples, enumerations etc.
+
+
+## Getting started
+
+
+To get started, first you need to understand that a ZIO-Schema is basically built-up from these three
+sealed traits:
+- `Record[R]`
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:102]: I would add `Primitive` here

File: docs/overview/understanding.md [hunk 8]
```diff
@@ -0,0 +1,328 @@
...
+type that has a `name` field, a `age` field, a `address` field, and a `friends` field.
+
+```scala
+case class User(name: String, age: Int, address: Address, friends: List[User])
+```
+
+This is called a "product type" in functional programming.
+The equivalent of a product type in ZIO-Schema is called a record.
+
+In ZIO-Schema such a record would be represented using the `Record[R]` type class:
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:157]: Nitpick: type class should be one word typeclass

**Review by thinkharderdev — COMMENTED** [review]
Thank you so much for this! Writing documentation is not the most glamorous work but it is super important and we're very grateful that you took this on! 

This looks great overall and I had a few small copy edit suggestions. One other thing is that to be consistent with documentation in other ZIO project I think we should refer to the library as "ZIO Schema" instead of "ZIO-Schema" (so drop the hyphen). It looks like that is the standard in other documentation. I would do the same when referring to other ZIO libraries here (so ZIO Flow instead of ZIO-Flow, etc). 

I also saw that you created a separate repo for examples. I think we should just fold that into the main repository in a zio-schema-examples subproject. 

Thanks!

File: docs/overview/first_schema.md [hunk 9]
```diff
@@ -0,0 +1,203 @@
+---
+id: first_schema
+title: "Your first ZIO Schema"
+---
+## Your First ZIO Schema
+
+ZIO Schema provides macros to help you create `Schema`s out of your data types. But before using the macros,
+we should take a look on how to do this the manual way.
```
Comment (TobiasPfeifer) [inline docs/overview/first_schema.md:8]: `take a look at` sounds more natural to me (non-native speaker)

File: docs/overview/first_schema.md [hunk 10]
```diff
@@ -0,0 +1,203 @@
...
+  }
+
+  final case class Customer(person: Person, paymentMethod: PaymentMethod)
+  
+}
+```
+
+### Manual construction of a Schema
+
+This part is similar to other libraries that you might now, e.g. for JSON processing.
```
Comment (TobiasPfeifer) [inline docs/overview/first_schema.md:31]: `now` = `know`

File: docs/overview/understanding.md [hunk 11]
```diff
@@ -0,0 +1,330 @@
...
+  def zip[B](that: Schema[B]): Schema[(A, B)] = Schema.Tuple(self, that)
+}
+```
+
+
+
+#### Records
+
+Your data structures usually are composed from a lot of types. For example, you might have a `User`
+type that has a `name` field, a `age` field, a `address` field, and a `friends` field.
```
Comment (TobiasPfeifer) [inline docs/overview/understanding.md:149]: a `age` field -> **an** `age` field
a `address` field -> **an** `address` field

File: docs/overview/understanding.md [hunk 12]
```diff
@@ -0,0 +1,330 @@
...
+    override def makeAccessors(b: AccessorBuilder): codec.Accessors[b.Lens, b.Prism, b.Traversal] =
+      codec.makeAccessors(b)
+
+    override def serializable: Schema[Schema[_]] = Meta(SchemaAst.fromSchema(codec))
+    override def toString: String                = s"Transform($codec)"
+  }
+```
+
+In the example above, we can transform the `User` Schema into a `UserRecord` Schema, which is a record,
+by using the `transform`-method, which has to be a "isomorphism" (= providing methods to transform A to B _and_ B to A):
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:289]: Copy edit: "has to be an isomorphism"

File: docs/overview/understanding.md [hunk 13]
```diff
@@ -0,0 +1,330 @@
...
+
+However, this is a "special language feature", it's not "real" like the side effects we've seen in the previous example ( `println(..) vs. Task`println(...)))` ).
+
+Because these basic operations are not "real", we're unable to create an operator that we can use to
+e.g. combine two fields that are inside a nested structure.
+
+The solution to this kind of problem is called an "Optic". Optics provide a way to access the fields of a case class and nested structures.
+There are three main types of optics:
+- `Lens`: A lens is a way to access a field of a case class.
+- `Prism`: A prim is a way to access a field of a nested structure or a collection.
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:57]: Copy edit: "A prism is a way to"

File: docs/overview/understanding.md [hunk 14]
```diff
@@ -0,0 +1,330 @@
...
+  sealed trait PaymentMethod 
+
+  object PaymentMethod {
+    final case class CreditCard(number: String, expirationMonth: Int, expirationYear: Int) extends PaymentMethod
+    final case class WireTransfer(accountNumber: String, bankCode: String) extends PaymentMethod
+  }
+```
+
+In functional programming, this kind of type is called a "sum type".
+In Scala3, this is called an "enum".
```
Comment (thinkharderdev) [inline docs/overview/understanding.md:185]: Suggestion:
```
In Scala2, this is called a "sealed trait"
In Scala3, this is called an "enum"
```

**Review by thinkharderdev — COMMENTED** [review]
A couple of small copy edits. Sorry, I had made a few comments before but forgot to submit the review :(

**Comment by domdorn** [comment]
updated the PR. If possible, lets merge this and then improve on it further with new PRs.

**Comment by thinkharderdev** [comment]
Looks great! Was there supposed to be a `zio-schema-examples` folder in this PR?

**Comment by domdorn** [comment]
yeah, haven't had time yet to create these - and not sure on how they should look like quality wise.. if you want, I can add those that I created in github.com/domdorn/zio-schema-example

**Comment by domdorn** [comment]
I've integrated the samples repository of @calvinlfer and me into this PR.. now github complains about files I haven't even touched.. :-/

**Review by thinkharderdev — APPROVED** [review]
A couple small comments but looks great! Still a conflict in the build.sbt but once it is resolved I think we're good.

File: zio-schema-examples/build.sbt [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+  val zioOpticsVersion = "0.1.0"
+
+  Seq(
+    zio %% "zio-schema"          % zioSchemaVersion,
+    zio %% "zio-schema-optics"   % zioSchemaVersion,
+    zio %% "zio-schema-protobuf" % zioSchemaVersion,
+    zio %% "zio-schema-json"     % zioSchemaVersion,
+    zio %% "zio-optics"          % zioOpticsVersion
+  )
+}
```
[the remark this answers is not in the document]
Comment (domdorn) [reply zio-schema-examples/build.sbt]: removed.

**Comment by domdorn** [comment]
I'm sorry to have to be so blunt: Work as well as private life is blowing up around myself at the moment.
I suggest to either merge this PR and then make any changes you seem appropriate or fork it and make a new PR or any other way you seem fit as I don't have time to make further adjustments to the PR in at least the next couple of weeks. I hope you understand my situation - I just hope the work done here doesn't go to waste, but I can't put any additional ressources into this at the moment.

**Comment by thinkharderdev** [comment]
> I'm sorry to have to be so blunt: Work as well as private life is blowing up around myself at the moment. I suggest to either merge this PR and then make any changes you seem appropriate or fork it and make a new PR or any other way you seem fit as I don't have time to make further adjustments to the PR in at least the next couple of weeks. I hope you understand my situation - I just hope the work done here doesn't go to waste, but I can't put any additional ressources into this at the moment.

Sorry to hear that @domdorn . I'll take care of getting the CI job fixed and get this merged. Thanks for your work on this!

**Comment by thinkharderdev** [comment]
Subsumed into #158
